### PR TITLE
[SYCL] Add prototype of ExtendedAtomics features

### PR DESCRIPTION
**Pull request by Pennycook**
Features:
- atomic_ref with integral and floating-point specializations
- atomic_fence

Tests:
- exchange
- compare_exchange
- fetch_add, +=, ++
- fetch_sub, -=, --
- fetch_min
- fetch_max

Signed-off-by: John Pennycook <john.pennycook@intel.com>
Co-authored-by: Alexey Sachkov <alexey.sachkov@intel.com>
Co-authored-by: Roland Schulz <roland.schulz@intel.com>

### CODE REVIEW

**Comment by Pennycook** [comment]
@bader What version of clang-format should I be using now? Is there a way to check that I'm using the right one?  My local version (built from intel/llvm) doesn't format the same way as clang-format-patch.

**Comment by bader** [comment]
CI check is using [clang-format-9](https://github.com/intel/llvm/blob/sycl/.github/workflows/clang-format.yml#L17).

**Comment by Pennycook** [comment]
> CI check is using [clang-format-9](https://github.com/intel/llvm/blob/sycl/.github/workflows/clang-format.yml#L17).

Perfect, thanks!

Two of the atomic tests are failing with the CUDA backend right now, so I've marked them as XFAIL.

**Comment by Pennycook** [comment]
We need more internal investigation to work out why the accelerator target is failing.  I've removed it from the tests for now and marked the extension as CPU- and GPU- only, to unblock this PR and enable people to experiment with it on the systems where it works as intended.

**Comment by bader** [comment]
@jbrodman, @mkinsner, @AlexeySachkov, @intel/llvm-reviewers-runtime, please, review.

File: sycl/include/CL/sycl/detail/spirv.hpp [hunk 1]
```diff
@@ -69,6 +70,221 @@ T GroupBroadcast(T x, id<Dimensions> local_id) {
...
+template <typename T, access::address_space AddressSpace>
+inline typename detail::enable_if_t<std::is_floating_point<T>::value, T>
+AtomicCompareExchange(multi_ptr<T, AddressSpace> mptr,
+                      intel::memory_scope scope, intel::memory_order success,
+                      intel::memory_order failure, T desired, T expected) {
+  using I = detail::make_unsinged_integer_t<T>;
+  auto spirv_success = getMemorySemanticsMask(success);
+  auto spirv_failure = getMemorySemanticsMask(failure);
+  auto spirv_scope = getScope(scope);
+  auto *ptr_int =
```
Comment (againull) [inline sycl/include/CL/sycl/detail/spirv.hpp:133]: Nice to have: follow llvm naming convention in the whole patch.
Ex, ptr_int -> PtrInt

File: sycl/include/CL/sycl/detail/spirv.hpp [hunk 2]
```diff
@@ -69,6 +70,221 @@ T GroupBroadcast(T x, id<Dimensions> local_id) {
...
+  case intel::memory_scope::device:
+    return __spv::Scope::Device;
+  case intel::memory_scope::system:
+    return __spv::Scope::CrossDevice;
+  }
+}
+
+template <typename T, access::address_space AddressSpace>
+inline typename detail::enable_if_t<std::is_integral<T>::value, T>
+AtomicCompareExchange(multi_ptr<T, AddressSpace> mptr,
```
Comment (againull) [inline sycl/include/CL/sycl/detail/spirv.hpp:113]: I am not sure, but probably we need to  disable all Atomic* templates for constant and private address space?

**Comment by againull** [comment]
@Pennycook , I am very sorry for being so late with review.

File: sycl/include/CL/sycl/detail/spirv.hpp [hunk 3]
```diff
@@ -69,6 +70,221 @@ T GroupBroadcast(T x, id<Dimensions> local_id) {
   return __spirv_GroupBroadcast(group_scope<Group>::value, ocl_x, ocl_id);
 }
 
+// Single happens-before means semantics should always apply to all spaces
+static inline constexpr __spv::MemorySemanticsMask::Flag
+getMemorySemanticsMask(intel::memory_order order) {
+  __spv::MemorySemanticsMask::Flag spv_order = __spv::MemorySemanticsMask::None;
+  switch (order) {
```
[the remark this answers is not in the document]
Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp:77]: Actually, I think this is a bug that was introduced in 4b6f467.  Each case originally contained a `return`, and I forgot to add the `break`s.  Only one of the cases is intended to fall through, and I'll add a comment for that one.  Good catch.

File: sycl/include/CL/sycl/detail/spirv.hpp (hunk 2 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp:113]: You're right that those address spaces aren't supported, and the user-facing API (`atomic_ref`) performs this check.  Do you think we need to add the checks to the implementation as well?  These functions should only be called inside `atomic_ref`.

File: sycl/include/CL/sycl/intel/atomic_enums.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+__SYCL_INLINE_CONSTEXPR memory_scope memory_scope_sub_group =
+    memory_scope::sub_group;
+__SYCL_INLINE_CONSTEXPR memory_scope memory_scope_work_group =
+    memory_scope::work_group;
+__SYCL_INLINE_CONSTEXPR memory_scope memory_scope_device = memory_scope::device;
+__SYCL_INLINE_CONSTEXPR memory_scope memory_scope_system = memory_scope::system;
+
+#ifndef __SYCL_DEVICE_ONLY__
+namespace detail {
+// Cannot use switch statement in constexpr before C++14
```
[the remark this answers is not in the document]
Comment (Pennycook) [reply sycl/include/CL/sycl/intel/atomic_enums.hpp:62]: Did this change recently?  I only added all of these things to satisfy the CI checks.

File: sycl/include/CL/sycl/detail/spirv.hpp (hunk 1 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp]: Fixed in 21bb880.  I followed the LLVM naming convention for all local variables, and for most function arguments.  I left names in snake_case in cases where the definition appears in a specification (e.g. C++, SYCL 1.2.1 or the ExtendedAtomics extension).

File: sycl/include/CL/sycl/detail/spirv.hpp (hunk 2 above)
Comment (againull) [reply sycl/include/CL/sycl/detail/spirv.hpp]: I see, check in the user-facing API looks enough, thanks!

File: sycl/include/CL/sycl/detail/spirv.hpp (hunk 1 above)
Comment (againull) [reply sycl/include/CL/sycl/detail/spirv.hpp]: Thanks!

File: sycl/include/CL/sycl/intel/atomic_enums.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (againull) [reply sycl/include/CL/sycl/intel/atomic_enums.hpp:62]: I see, this code is not part of the library, but goes with headers, so looks like both versions are needed.

[the remark this answers is not in the document]
Comment (Pennycook) [reply sycl/include/CL/sycl/intel/atomic_enums.hpp:62]: Ah, okay.  Yes, I think this needs to go in the headers.  Thanks for checking.

File: sycl/include/CL/__spirv/spirv_ops.hpp [hunk 5]
```diff
@@ -155,6 +155,7 @@ extern SYCL_EXTERNAL TempRetT __spirv_ImageSampleExplicitLod(SampledType,
       macro(__attribute__((opencl_local)), Arg)
 
 __SPIRV_ATOMICS(__SPIRV_ATOMIC_FLOAT, float)
+__SPIRV_ATOMICS(__SPIRV_ATOMIC_FLOAT, double)
```
Comment (v-klochkov) [inline sycl/include/CL/__spirv/spirv_ops.hpp:158]: Just a question: Is 'half' type expected to be supported too eventually?

Comment (Pennycook) [reply sycl/include/CL/__spirv/spirv_ops.hpp:158]: Good question.  OpenCL only defines atomics for 32- and 64-bit types, which is what I've started with here.

C++ allows an `atomic_ref` to be constructed from any `TriviallyCopyable` type, and I think eventually that's where we'd like to end up.  Whether the `atomic_ref` for a given type is implemented on top of native instructions, a compare-exchange loop, a global lock, etc, would then be implementation-defined and device-specific.

We haven't yet defined exactly what sort of device queries we'll want to support for testing atomic functionality.  If you have any ideas for what the feature sets should look like, or if there are any queries that would be useful for implementing reductions, please let me know!

**Comment by v-klochkov** [comment]
This PR requires approval from @AlexeySachkov and @mkinsner as code-owners.

**Comment by jbrodman** [comment]
I should be able to approve for @mkinsner.

**Comment by jbrodman** [comment]
@bader - how can we get this out of review hell?

**Comment by bader** [comment]
@AlexeySachkov, could you take a look, please?

**Comment by jbrodman** [comment]
Thank you. Please merge ASAP.

**Comment by bader** [comment]
> Thank you. Please merge ASAP.

I think we should run the tests again. This PR was tested 9 days ago. I'd to make sure it still passes the tests.
I'll try to re-run the tests and merge if they pass.

**Comment by jbrodman** [comment]
> > Thank you. Please merge ASAP.
> 
> I think we should run the tests again. This PR was tested 9 days ago. I'd to make sure it still passes the tests.
> I'll try to re-run the tests and merge if they pass.

Makes sense.

**Pull request merged**